govcd/media.go

```diff
@@ -354,3 +413,220 @@ func (mediaItem *MediaItem) Refresh() error {
 
 	return err
 }
+
+// Refresh refreshes the media information by href
+func (media *Media) Refresh() error {
+
+	if media.Media == nil {
+		return fmt.Errorf("cannot refresh, Object is empty")
+	}
+
+	url := media.Media.HREF
+
+	// Empty struct before a new unmarshal, otherwise we end up with duplicate
+	// elements in slices.
+	media.Media = &types.Media{}
+
+	_, err := media.client.ExecuteRequest(url, http.MethodGet,
+		"", "error retrieving media: %s", nil, media.Media)
+
+	return err
+}
+
+// GetMediaByHref finds a Media by HREF
+// On success, returns a pointer to the Media structure and a nil error
+// On failure, returns a nil pointer and an error
+func (cat *Catalog) GetMediaByHref(mediaHref string) (*Media, error) {
+
+	media := NewMedia(cat.client)
+
+	_, err := cat.client.ExecuteRequest(mediaHref, http.MethodGet,
+		"", "error retrieving media: %s", nil, media.Media)
+	if err != nil {
+		return nil, err
+	}
+	return media, nil
+}
+
+// GetMediaByName finds a Media by Name
+// On success, returns a pointer to the Media structure and a nil error
+// On failure, returns a nil pointer and an error
+func (cat *Catalog) GetMediaByName(mediaName string, refresh bool) (*Media, error) {
+	if refresh {
+		err := cat.Refresh()
+		if err != nil {
+			return nil, err
+		}
+	}
+	for _, catalogItems := range cat.Catalog.CatalogItems {
+		for _, catalogItem := range catalogItems.CatalogItem {
+			if catalogItem.Name == mediaName && catalogItem.Type == "application/vnd.vmware.vcloud.catalogItem+xml" {
+				catalogItemElement, err := cat.GetCatalogItemByHref(catalogItem.HREF)
+				if err != nil {
+					return nil, err
+				}
+				return cat.GetMediaByHref(catalogItemElement.CatalogItem.Entity.HREF)
+			}
+		}
+	}
+	return nil, ErrorEntityNotFound
+}
+
+// GetMediaById finds a Media by ID
+// On success, returns a pointer to the Media structure and a nil error
+// On failure, returns a nil pointer and an error
+func (cat *Catalog) GetMediaById(mediaId string, refresh bool) (*Media, error) {
+	if refresh {
+		err := cat.Refresh()
+		if err != nil {
+			return nil, err
+		}
+	}
+	mediaHREF := cat.client.VCDHREF
+
+	mediaBareId, err := getBareEntityUuid(mediaId)
+	if err != nil {
+		util.Logger.Printf("[Error] parsing bareID from mediaId %s: %s", mediaId, err)
+		return nil, ErrorEntityNotFound
```

Comment:
simply return the error received from getBareEntityUuid

Reply:
tx, changed

Reply:
It was not. Recent commit started returning `ErrorEntityNotFound` on parsing error:
https://github.com/vbauzysvmware/go-vcloud-director/blob/changes-for-media/govcd/media.go#L487

We discussed at the beginning that this is not right.

Reply:
It is a bad practice to assume knowledge about the calling functions from within a different function. Such assumptions lead to bad design choices and poorly factorized code. This is especially import when dealing with public functions.

`GetById` returns `cat.GetMediaByHref(mediaHREF.String())`. `GetMediaByHref` has been refactored to return `ErrorEntityNotFound`, hence returning of `ErrorEntityNotFound` on parsing error shouldn't be needed.

Reply:
The error that we get here is "the parsing went wrong". We can't return any "not found".
GetMediaByNameOrId assumes we have implemented things according to the same principles.
If you want to go ahead with this implementation, please take that into account.

Reply:
After further investigation (we really need to converge)...

@dataclouder the generic `getEntityByNameOrId` is a "blind" function - it _doesn't understand_ whether it is given an ID or name and tries to call `getById` first and then `getByName`. I realize that we need this function for backwards compatibility when switching identifier from name to ID, but now it starts to bite:

```
	entity, byIdErr = getById(identifier, refresh)
	if byIdErr == nil {
		// Found by ID
		return entity, nil
	}
	if IsNotFound(byIdErr) {
		// Not found by ID, try by name
		entity, byNameErr = getByName(identifier, false)
		return entity, byNameErr
	} else {
		// On any other error, we return it
		return nil, byIdErr
	}
```

If `GetMediaById` is provided with a wrong ID (i.e. name) we could make it:
a. Return a generic parsing error. This is correct behavior from `GetMediaById` perspective, but then we need to fix the generic `getEntityByNameOrId`, because now `getEntityByNameOrId` would **not** try to search by `getByName` as it is only doing that if `if IsNotFound(byIdErr)` (!) I personally don't see how to make `getEntityByNameOrId` smarter at this point, because even the user could use name in such format like our IDs.
b. Return `ErrorEntityNotFound`. This is not ideal, but after more thought an incorrectly provided ID (like a name instead of ID) does imply it won't be found, so we're only one logic step ahead. Then this complies with what the generic `getEntityByNameOrId` expects.
c. Introduce and return a specific `ErrorParsing` and then check for that in the generic `getEntityByNameOrId` function.

Reply:
We agreed that we'll try one more variant before going the (b) or (c) ways:

d. Try to use `queryMediaItemsWithFilter` with catalog filter and loop/match with the help of `equalIds(...)`.

Risk: query sometimes doesn't return results without a clear reason, we have seen that in some cases with filtering PVDCs.